### PR TITLE
Response output may not close at end of dispatch forward

### LINKED ISSUE #16053: Can not close response output stream at the end of dispatch forward
**Describe the bug** 
When the server's response is wrapped and used in the dispatch forward, the response's output stream is not closed before exiting the forward.

**Steps to Reproduce**  
Wrap the response object then dispatch forward.  Upon returning from forward, write data into the response.  These extra data after dispatch forward will be seen by the client.

**Expected behavior**  
The response's outputStream is closed before exiting dispatch forward.

**Diagnostic information:**  
- OpenLiberty: any
-- Affected feature(s) [servlet]
 
**Additional context**  
In general, there is no disruption in the flow or any exception.  Client receives normal response data.  

In a case where an application continues writing after the forward, additional data will be written into the response (since the response output stream is not closed).  In that case, these extra data are seen by the client.  This is probably the only symptom observed from the client.

### PR DESCRIPTION
**Pull request by pmd1nh**
fixes #16053

### CODE REVIEW

**Comment by pmd1nh** [comment]
#build       #spawn.fullfat.buckets=com.ibm.ws.webcontainer_fat,com.ibm.ws.webcontainer.servlet.3.1_fat,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.webcontainer_fat_lWAS,com.ibm.ws.webcontainer_fat_servlet31_webcontainerlWAS,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0.lWAS

**Comment by pmd1nh** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.webcontainer_fat,com.ibm.ws.webcontainer.servlet.3.1_fat,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.webcontainer_fat_lWAS,com.ibm.ws.webcontainer_fat_servlet31_webcontainerlWAS,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0.lWAS

**Comment by pmd1nh** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.webcontainer_fat,com.ibm.ws.webcontainer.servlet.3.1_fat,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.webcontainer_fat_lWAS,com.ibm.ws.webcontainer_fat_servlet31_webcontainerlWAS,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0.lWAS

**Comment by pmd1nh** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.webcontainer_fat,com.ibm.ws.webcontainer.servlet.3.1_fat,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.webcontainer_fat_lWAS,com.ibm.ws.webcontainer_fat_servlet31_webcontainerlWAS,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0.lWAS

**Comment by pmd1nh** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.webcontainer_fat,com.ibm.ws.webcontainer.servlet.3.1_fat,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.webcontainer_fat_lWAS,com.ibm.ws.webcontainer_fat_servlet31_webcontainerlWAS,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0.lWAS

**Comment by pmd1nh** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.webcontainer_fat,com.ibm.ws.webcontainer.servlet.3.1_fat,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.webcontainer_fat_lWAS,com.ibm.ws.webcontainer_fat_servlet31_webcontainerlWAS,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0.lWAS

**Comment by pmd1nh** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.webcontainer_fat,com.ibm.ws.webcontainer.servlet.3.1_fat,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.webcontainer_fat_lWAS,com.ibm.ws.webcontainer_fat_servlet31_webcontainerlWAS,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0.lWAS

**Comment by pmd1nh** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.webcontainer_fat,com.ibm.ws.webcontainer.servlet.3.1_fat,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.webcontainer_fat_lWAS,com.ibm.ws.webcontainer_fat_servlet31_webcontainerlWAS,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0.lWAS

**Comment by pmd1nh** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.webcontainer_fat,com.ibm.ws.webcontainer.servlet.3.1_fat,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.webcontainer_fat_lWAS

**Comment by pmd1nh** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.webcontainer_fat,com.ibm.ws.webcontainer.servlet.3.1_fat,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.webcontainer_fat_lWAS

**Comment by pmd1nh** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.webcontainer_fat,com.ibm.ws.webcontainer.servlet.3.1_fat,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.webcontainer_fat_lWAS

**Comment by pmd1nh** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.webcontainer_fat,com.ibm.ws.webcontainer.servlet.3.1_fat,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.webcontainer_fat_lWAS

**Comment by pmd1nh** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.webcontainer_fat,com.ibm.ws.webcontainer.servlet.3.1_fat,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.webcontainer_fat_lWAS

**Comment by pmd1nh** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.webcontainer_fat,com.ibm.ws.webcontainer.servlet.3.1_fat,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.webcontainer_fat_lWAS

**Comment by pmd1nh** [comment]
#build

File: dev/com.ibm.ws.webcontainer/resources/OSGI-INF/l10n/metatype.properties [hunk 1]
```diff
@@ -185,4 +185,7 @@ setHtmlContentTypeOnError.name=Disable text/html content type on error responses
...
 
 excludeAllHandledTypesClasses.name=Disable returning handlesTypes parameters
-excludeAllHandledTypesClasses.desc=When this property is set to true, during startup, ServletContainerInitializer implementors that use the HandlesTypes annotation do not receive classes that are specified as HandlesTypes parameters.
\ No newline at end of file
+excludeAllHandledTypesClasses.desc=When this property is set to true, during startup, ServletContainerInitializer implementors that use the HandlesTypes annotation do not receive classes that are specified as HandlesTypes parameters.
+
+closeWrappedResponseOutputAfterForward.name=Close wrapped response output after forward
+closeWrappedResponseOutputAfterForward.desc=When this property is set to true, the wrapped response's output stream will be closed before exiting the dispatch forward.  The default value is false. For servlet-5.0 and higher, the default value is true.
```
Comment (dmuelle) [inline dev/com.ibm.ws.webcontainer/resources/OSGI-INF/l10n/metatype.properties]: When this property is set to true, the wrapped response's output stream will be closed before exiting the dispatch forward
--->
When this property is set to true, the output stream from the wrapped response is closed before it exits the dispatch forward.

File: dev/com.ibm.ws.webcontainer/resources/OSGI-INF/l10n/metatype.properties [hunk 2]
```diff
@@ -185,4 +185,7 @@ setHtmlContentTypeOnError.name=Disable text/html content type on error responses
 setHtmlContentTypeOnError.desc=When this property is false, the webcontainer will not set the response's content type header during the error handling process. An application is responsible to set the response's content type. The default value (true) sets the content type to "text/html".
 
 excludeAllHandledTypesClasses.name=Disable returning handlesTypes parameters
```
Comment (dmuelle) [inline dev/com.ibm.ws.webcontainer/resources/OSGI-INF/l10n/metatype.properties:187]: Disable returning handlesTypes parameters
--->
Disable the return of handlesTypes parameters

File: dev/com.ibm.ws.webcontainer/resources/OSGI-INF/l10n/metatype.properties [hunk 3]
```diff
@@ -185,4 +185,7 @@ setHtmlContentTypeOnError.name=Disable text/html content type on error responses
...
-excludeAllHandledTypesClasses.desc=When this property is set to true, during startup, ServletContainerInitializer implementors that use the HandlesTypes annotation do not receive classes that are specified as HandlesTypes parameters.
\ No newline at end of file
+excludeAllHandledTypesClasses.desc=When this property is set to true, during startup, ServletContainerInitializer implementors that use the HandlesTypes annotation do not receive classes that are specified as HandlesTypes parameters.
+
+closeWrappedResponseOutputAfterForward.name=Close wrapped response output after forward
```
Comment (dmuelle) [inline dev/com.ibm.ws.webcontainer/resources/OSGI-INF/l10n/metatype.properties]: Close wrapped response output after forward
--->
Close wrapped-response output after forward

**Review by dmuelle — CHANGES_REQUESTED** [review]
a few minor suggestions

**Review by dmuelle — APPROVED** [review]
LGTM

**Comment by pmd1nh** [comment]
#build       #spawn.fullfat.buckets=com.ibm.ws.webcontainer_fat,com.ibm.ws.webcontainer.servlet.3.1_fat,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.webcontainer_fat_part2,com.ibm.ws.webcontainer_fat_servlet31toleration,com.ibm.ws.webcontainer_fat_servlet31toleration_part2,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0_part2,com.ibm.ws.webcontainer_fat_lWAS,com.ibm.ws.webcontainer_fat_servlet31_webcontainerlWAS,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0.lWAS

**Comment by pmd1nh** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.webcontainer_fat,com.ibm.ws.webcontainer.servlet.3.1_fat,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.webcontainer_fat_part2,com.ibm.ws.webcontainer_fat_servlet31toleration,com.ibm.ws.webcontainer_fat_servlet31toleration_part2,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0_part2,com.ibm.ws.webcontainer_fat_lWAS,com.ibm.ws.webcontainer_fat_servlet31_webcontainerlWAS,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0.lWAS

**Comment by pmd1nh** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.webcontainer_fat,com.ibm.ws.webcontainer.servlet.3.1_fat,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.webcontainer_fat_part2,com.ibm.ws.webcontainer_fat_servlet31toleration,com.ibm.ws.webcontainer_fat_servlet31toleration_part2,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0_part2,com.ibm.ws.webcontainer_fat_lWAS,com.ibm.ws.webcontainer_fat_servlet31_webcontainerlWAS,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0.lWAS

**Comment by pmd1nh** [comment]
#build

**Comment by pmd1nh** [comment]
#build

**Comment by pmd1nh** [comment]
#build 
#spawn.fullfat.buckets=com.ibm.ws.webcontainer.servlet.4.0_fat

**Comment by pmd1nh** [comment]
#build
#spawn.fullfat.buckets=com.ibm.ws.webcontainer.servlet.4.0_fat

**Comment by pmd1nh** [comment]
Updated test application

**Comment by pmd1nh** [comment]
#build
#spawn.fullfat.buckets=com.ibm.ws.webcontainer.servlet.4.0_fat

**Comment by pmd1nh** [comment]
#build
#spawn.fullfat.buckets=com.ibm.ws.webcontainer.servlet.4.0_fat

**Comment by pmd1nh** [comment]
test_nonWrapResponse_closeOutputStreamAfterForward | 1 | 0 | 0 | 1 s | 1
-- | -- | -- | -- | -- | --
test_wrapResponse_closeOutputStreamAfterForward | 1 | 0 | 0 | 0 s | 2
test_nonWrapResponse_closeOutputStreamAfterForward_EE9_FEATURES | 1 | 0 | 0 | 1 s | 3
test_wrapResponse_closeOutputStreamAfterForward_EE9_FEATURES | 1 | 0 | 0 | 0 s | 4

File: dev/com.ibm.ws.webcontainer.servlet.4.0_fat/test-applications/CloseWrappedResponseOutputForward.war/resources/WEB-INF/web.xml [hunk 4]
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (pnicolucci) [inline dev/com.ibm.ws.webcontainer.servlet.4.0_fat/test-applications/CloseWrappedResponseOutputForward.war/resources/WEB-INF/web.xml:1]: Add copyright header.

Comment (pnicolucci) [inline dev/com.ibm.ws.webcontainer.servlet.4.0_fat/test-applications/CloseWrappedResponseOutputForward.war/resources/WEB-INF/web.xml:1]: Add copyright header

File: dev/com.ibm.ws.webcontainer/resources/OSGI-INF/l10n/metatype.properties [hunk 5]
```diff
@@ -185,4 +185,7 @@ setHtmlContentTypeOnError.name=Disable text/html content type on error responses
...
 
 excludeAllHandledTypesClasses.name=Disable returning handlesTypes parameters
-excludeAllHandledTypesClasses.desc=When this property is set to true, during startup, ServletContainerInitializer implementors that use the HandlesTypes annotation do not receive classes that are specified as HandlesTypes parameters.
\ No newline at end of file
+excludeAllHandledTypesClasses.desc=When this property is set to true, during startup, ServletContainerInitializer implementors that use the HandlesTypes annotation do not receive classes that are specified as HandlesTypes parameters.
+
+closeWrappedResponseOutputAfterForward.name=Close wrapped-response output after forward
+closeWrappedResponseOutputAfterForward.desc=When this property is set to true, the output stream from the wrapped response is closed before it exits the dispatch forward. The default value is false. For servlet-5.0 and higher, the default value is true.
```
Comment (pnicolucci) [inline dev/com.ibm.ws.webcontainer/resources/OSGI-INF/l10n/metatype.properties]: Can you just add a newline here to satisfy git?

File: dev/com.ibm.ws.webcontainer.servlet.4.0_fat/fat/src/com/ibm/ws/fat/wc/tests/WCCloseWrappedResponseOutputAfterForwardTest.java [hunk 6]
```diff
@@ -0,0 +1,124 @@
+/*******************************************************************************
```
Comment (pnicolucci) [inline dev/com.ibm.ws.webcontainer.servlet.4.0_fat/fat/src/com/ibm/ws/fat/wc/tests/WCCloseWrappedResponseOutputAfterForwardTest.java:1]: My understanding of this test is you're only testing the `true` scenario for all servlet versions. Can we also test the default behavior for servlet 5.0 and default value for previous servlet versions?

File: dev/com.ibm.ws.webcontainer/src/com/ibm/wsspi/webcontainer/WCCustomProperties.java [hunk 7]
```diff
@@ -321,9 +321,12 @@
 
     //21.0.0.4
     public static boolean SET_HTML_CONTENT_TYPE_ON_ERROR; 
-
+    
     //21.0.0.6
     public static boolean EXCLUDE_ALL_HANDLED_TYPES_CLASSES;
+    
+    //21.0.0.11
```
Comment (pnicolucci) [inline dev/com.ibm.ws.webcontainer/src/com/ibm/wsspi/webcontainer/WCCustomProperties.java]: Also this release isn't correct.

File: dev/com.ibm.ws.webcontainer/src/com/ibm/wsspi/webcontainer/WCCustomProperties.java [hunk 8]
```diff
@@ -804,31 +808,29 @@ public static void setCustomPropertyVariables() {
...
             ALLOW_QUERY_PARAM_WITH_NO_EQUAL = Boolean.valueOf(WebContainer.getWebContainerProperties().getProperty("com.ibm.ws.webcontainer.allowqueryparamwithnoequal", "true")).booleanValue();
-            //21.0.0.6
             EXCLUDE_ALL_HANDLED_TYPES_CLASSES = Boolean.valueOf(WebContainer.getWebContainerProperties().getProperty("com.ibm.ws.webcontainer.excludeallhandledtypesclasses", "true")).booleanValue();
+
             //21.0.0.10
             ENABLE_POST_ONLY_J_SECURITY_CHECK = Boolean.valueOf(WebContainer.getWebContainerProperties().getProperty("com.ibm.ws.webcontainer.enablepostonlyjsecuritycheck", "true")).booleanValue(); //18368
             //21.0.0.11
             DEFER_SERVLET_REQUEST_LISTENER_DESTROY_ON_ERROR =  Boolean.valueOf(WebContainer.getWebContainerProperties().getProperty("com.ibm.ws.webcontainer.deferservletrequestlistenerdestroyonerror", "true")).booleanValue(); //PI26908
-
+            CLOSE_WRAPPED_RESPONSE_OUTPUT_AFTER_FORWARD = Boolean.valueOf(WebContainer.getWebContainerProperties().getProperty("com.ibm.ws.webcontainer.closewrappedresponseoutputafterforward", "true")).booleanValue();
```
Comment (pnicolucci) [inline dev/com.ibm.ws.webcontainer/src/com/ibm/wsspi/webcontainer/WCCustomProperties.java:825]: This won't go out in 21.0.0.11 so we need to update.

**Review by pnicolucci — APPROVED** [review]

**Comment by pmd1nh** [comment]
#build       #spawn.fullfat.buckets=com.ibm.ws.webcontainer_fat,com.ibm.ws.webcontainer.servlet.3.1_fat,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.webcontainer_fat_part2,com.ibm.ws.webcontainer_fat_servlet31toleration,com.ibm.ws.webcontainer_fat_servlet31toleration_part2,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0_part2,com.ibm.ws.webcontainer_fat_lWAS,com.ibm.ws.webcontainer_fat_servlet31_webcontainerlWAS,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0.lWAS

**Comment by pmd1nh** [comment]
Tests: |   | 84435 tests, 0 failures, 1 errors
-- | -- | --

one unrelated error in:

com.ibm.ws.security.wim.registry_fat FAT tests | 59 | 0 | 1 |  
-- | -- | -- | -- | --




com.ibm.ws.fat.wc.tests.WC5CloseWrappedResponseOutputAfterForwardTest | 2 | 0 | 0 | 2 s | 26
-- | -- | -- | -- | -- | --
test_50nonWrapResponse_closeOutputStreamAfterForward_EE9_FEATURES | 1 | 0 | 0 | 2 s | 1
test_50wrapResponse_closeOutputStreamAfterForward_EE9_FEATURES | 1 | 0 | 0 | 0 s | 2
 



com.ibm.ws.fat.wc.tests.WCCloseWrappedResponseOutputAfterForwardTest | 4 | 0 | 0 | 3 s | 23
-- | -- | -- | -- | -- | --
test_nonWrapResponse_closeOutputStreamAfterForward | 1 | 0 | 0 | 1 s | 1
test_wrapResponse_closeOutputStreamAfterForward | 1 | 0 | 0 | 0 s | 2
test_nonWrapResponse_closeOutputStreamAfterForward_EE9_FEATURES | 1 | 0 | 0 | 1 s | 3
test_wrapResponse_closeOutputStreamAfterForward_EE9_FEATURES | 1 | 0 | 0 | 0 s | 4

**Comment by pmd1nh** [comment]
#run-libby-bot

**Pull request merged**